### ISSUE TITLE
Colormap artifact in mipmaps along region boundaries

### ISSUE THREAD

**Issue by TokisanGames** [TokisanGames/Terrain3D#284]
### Terrain3D version

v0.9.0

### System information

Godot v4.1.3.stable - Windows 10.0.22621 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3070 Laptop GPU (NVIDIA; 31.0.15.4633) - 12th Gen Intel(R) Core(TM) i9-12900H (20 Threads)

### Is the issue reproducable in the demo?

Yes

### Issue description

Mipmaps were generated on the colormap array in 7e4f9ebe985236ba0a45d13581847193b46a4eb1 to avoid distant noise, which was very obvious when using a full colormap such as in GIS applications w/ satellite photos (farther below). However, this has introduced another artifact on the region boundary. Enabling `renormalize` on `generate_mipmaps()` has no affect.

`ALBEDO=colormap*vec3(.1)`

![image](https://github.com/TokisanGames/Terrain3D/assets/632766/08a44db7-d5a1-46eb-8dc4-c969f8464593)

Regular shader

![image](https://github.com/TokisanGames/Terrain3D/assets/632766/c13f44ae-841b-4579-a119-514cee418d0e)

Though mipmaps should be generated on the whole map at once, it appears this tiling is not a problem and the individual mipmaps blend together fine. It's actually surprising that they blend at all, given how different from each other the lower lods look. 

![image](https://github.com/TokisanGames/Terrain3D/assets/632766/7ec706aa-3fc7-4b18-b25c-aaa937bb05ab)

The only artifact is the small grid line that shows up when up close on mipmap0 shown at the start, and also far away shown here.

![image](https://github.com/TokisanGames/Terrain3D/assets/632766/ebc66d92-0ab0-4e7a-a178-2ed3d5fb2dea)

Upon further testing, I've found that indeed it is not the mipmap lods, it is actually the system that determines which LOD to use that is the problem.

![image](https://github.com/TokisanGames/Terrain3D/assets/632766/e3ef8053-ec66-49d1-806f-0be274ef81ef)

These lines are the borders of the individual textures in the arrays and show up when using `texture()`.

I have found I can clamp the lod based on distance.

![image](https://github.com/TokisanGames/Terrain3D/assets/632766/6cbc6e9c-0623-48ff-b21d-ce79c27074fd)

 I could go up to 2048 or 3172 to basically eliminate the grid at all distances. However in practice with either a full GIS colormap, or ground textures up close, 1024 gives me the best results of texture mipmaping to eliminate noise at far distances, and fix this line at close distances.

![image](https://github.com/TokisanGames/Terrain3D/assets/632766/39d5a102-0506-4989-a4c2-7465b233e6fc)

I'll close this issue with this workaround, though I believe this to be an engine bug as `textureQueryLod()` should not show a grid and should not bleed into my renders using `texture()`.

`textureLod(tex, 0)` does not include this grid artifact.

Basically these two lookup blocks provide very similar results. It's as if when `texture()` is used on texture arrays, it adds the results of `textureQueryLod()` to the result.

```
# Texture Lookup
	color_map = texture(_color_maps, region_uv);

# TextureLod + textureQueryLod artifically added
	vec2 v = textureQueryLod(_color_maps, region_uv.xy);
	color_map.rgb = textureLod(_color_maps, region_uv, v.y).rgb + vec3(v.y*.005);

ALBEDO = albedo_height.rgb * color_map.rgb;
```

Upstream issues
https://github.com/godotengine/godot/issues/48462

https://github.com/godotengine/godot/issues/27837#issuecomment-1134524516

https://github.com/godotengine/godot/issues/33519

**Comment by TokisanGames** [comment]
I reopened this because I think this is an upstream bug in `texture` and `textureQueryLod`.  https://github.com/TokisanGames/Terrain3D/commit/7860ecf3475f6f40896b9a3a51e3185bf48f7a74 is a workaround, but I don't think it should be there at all.

To test it: 
1. Download the [latest release](https://github.com/TokisanGames/Terrain3D/releases/tag/v0.9.0-beta) of terrain3d
2. Load it and restart twice
3. Open the demo
4. Hide the borders node
5. Enable the Material/Shader override enabled and paste this [minimal_shader.txt](https://github.com/TokisanGames/Terrain3D/files/13761309/minimal_shader.txt) in it. 

In the shader you'll see three sections highlighting:
1. the problem with `textureQueryLod` showing the texture array borders
2. the problem with `texture` seemingly combining the above values into the results
3. the current hacky workaround which as good as if it were working properly.

With (2) enabled, you'll see the artifact along the path at the region borders, shown in the red artifact below. The green artifact is unrelated, tracked in #185. 

![image](https://github.com/TokisanGames/Terrain3D/assets/632766/0d744510-1df5-4ec4-879e-7cbf825507bc)

**Comment by clayjohn** [comment]
@TokisanGames I intend to take a look a bit more closely later, but here are the first two things I will look for:
1. Since you are calculating UV2 by flooring the varying you calculate in the vertex shader you can easily fall prey to precision issues. Oftentimes 0.0s and 1.0s can end up slightly off. So when you `floor()` the result you end up with -1.0 (when you wanted 0.0) or 0.0 (when you wanted 1.0). Oftentimes using `round()` is safer, or just adding a very small bias.
2. Issues with the derivative `texture()` and `textureQueryLod()` rely on the derivative of the passed UV coordinate to calculate LOD level. If you UV is non-continous (i.e. it jumps at points) then the derivative may explode causing the function to suddenly use a very high mip level. That was the issue with https://github.com/godotengine/godot/issues/48462 where the user manually clamped the UV coordinate instead of using the repeat mode on the texture. 

My gut is leaning towards this being an issue with the derivative. When using custom UV coordinates that are computed in the shader you pretty much always have to calculate the derivative yourself and then use `textureLod()` or `textureGrad()`.

**Comment by TokisanGames** [comment]
Thanks for looking at this @clayjohn 

> Since you are calculating UV2 by flooring the varying you calculate in the vertex shader you can easily fall prey to precision issues. Oftentimes 0.0s and 1.0s can end up slightly off. 

Hmm, I understand. However here, if I change floor to round in get_region_uv/2 it breaks the terrain. Every other region is messed up.

> My gut is leaning towards this being an issue with the derivative. 
> When using custom UV coordinates that are computed in the shader you pretty much always have to calculate the derivative yourself and then use textureLod() or textureGrad().

I haven't studied derivatives in over 20 years, and never in the context of shaders. Do you have a link where I can learn what I need? I've found it hard to find anything that is not arcane phonebook information on textureQueryLod, texturelod, texturegrad, etc. 

We [used to use textureGrad](https://github.com/TokisanGames/Terrain3D/issues/169) (see update 3). But I found that using `texture` was 20% faster for the same result. I'd rather use the current hacky method of clamping `textureQueryLod` and using `textureLod` for the speed than pay 20%. Do you know where I can find the formula that `texture` uses to determine the lod level? Is there a non-slow way to use derivatives, or the same internal code of `texture`?

**Comment by clayjohn** [comment]
Ben Golus has a detailed medium article discussing ways to deal with this same issue https://bgolus.medium.com/distinctive-derivative-differences-cce38d36797b. I don't really have any definitive resources to learn about derivatives though. 

Manual derivatives tend to be slow compared to the automatically calculated ones as the automatically calculated ones are often done in hardware and/or with a bunch of GPU-specific tricks. So I'm not surprised that manually calculating them ended up being a bit slower. 20% seems like a lot, but I guess it depends on what your results are, if you end up with much lower LODs, the shader can become way more expensive. 

Basically what you need is a good way of approximating the per-pixel change in your UV channel without the discontinuities introduced by ``get_region_uv2()``. Since ``get_region_uv2()`` just offsets the uv value, you can just use the input value (``UV2``) instead. 

Accordingly, the easy workaround is to do the following:
```
vec3 region_uv = get_region_uv2(UV2);
float lod = textureQueryLod(_color_maps, UV2).y;
color_map = textureLod(_color_maps, region_uv, lod);
```

Testing locally I can confirm that in the "problem 1" case the artifacts go away when using UV2 instead of region_uv. However, I can't reproduce the issue when using any of the other "problems" or when using actual color.

![Screenshot from 2024-01-07 17-23-41](https://github.com/TokisanGames/Terrain3D/assets/16521339/44ecf433-de10-49a1-9017-b0b912c8dc50)

![Screenshot from 2024-01-07 17-23-50](https://github.com/TokisanGames/Terrain3D/assets/16521339/373829c5-af1b-477c-97f9-a094c42f0d17)

**Comment by TokisanGames** [comment]
@clayjohn I'm sorry I'm slow to respond. Thank you very much for looking at this and providing a solution. This works well.

Fixed in a096e6cc5bb4d639c474cbf8c353028462f44365
